test/topology_custom/test_raft_recovery_user_data.py

```diff
@@ -0,0 +1,192 @@
+    """
+    Test that the Raft-based recovery procedure works correctly with the user data. It involves testing:
+    - client requests during the procedure (mainly availability),
+    - removing/replacing dead nodes during the procedure in the presence of client requests and tablets on dead nodes.
+
+    1. Start a cluster with two dcs, dc1 and dc2, containing three nodes each.
+    2. Start sending writes with CL=LOCAL_QUORUM to a table with RF=3, NetworkTopologyStrategy and tablets.
+    3. Kill all nodes from dc2 causing a permanent group 0 majority loss.
+    4. Run the recovery procedure to recreate group 0 with nodes from dc2 as new members. Writes sent to dc1 should
+    continue succeeding since at least two nodes are alive at any point during the recovery procedure (it involves only
+    rolling restarts), every node in dc1 is a replica (3 nodes, RF=3) and two nodes make a local quorum in dc1.
+    5. Remove nodes from dc2 from topology using remove or replace, depending on the value of the remove_dead_nodes_with
+    parameter. For remove, we must do two additional steps to make it work:
+    - Mark all dead nodes as permanently dead.
+    - Decrease RF of the user keyspace, which uses tablets, to 0 in dc2.
+    6. For remove, add 3 nodes to dc2 and bring back RF of the user keyspace to 3 in dc2, which brings back dc2 to its
+    previous state. For replace, add a single node (a sanity check verifying that the cluster is functioning properly).
+    7. Stop sending writes.
+    """
+    # Decrease failure_detector_timeout_in_ms from the default 20 s to speed up some graceful shutdowns in the test.
+    # Shutting down the CQL server can hang for failure_detector_timeout_in_ms in the presence of dead nodes and
+    # CQL requests.
+    cfg = {
+        'endpoint_snitch': 'GossipingPropertyFileSnitch',
+        'enable_tablets': True,
+        'failure_detector_timeout_in_ms': 2000,
+    }
+    property_file_dc1 = {'dc': 'dc1', 'rack': 'rack1'}
+    property_file_dc2 = {'dc': 'dc2', 'rack': 'rack2'}
+
+    logging.info('Adding servers that will survive majority loss to dc1')
+    live_servers = await manager.servers_add(3, config=cfg, property_file=property_file_dc1)
+    logging.info('Adding servers that will be killed to dc2')
+    dead_servers = await manager.servers_add(3, config=cfg, property_file=property_file_dc2)
+    logging.info(f'Servers to survive majority loss: {live_servers}, servers to be killed: {dead_servers}')
+
+    cql, _ = await manager.get_ready_cql(live_servers + dead_servers)
+    hosts = await wait_for_cql_and_get_hosts(cql, live_servers, time.time() + 60)
+
+    first_group0_id = (await cql.run_async(
+            "SELECT value FROM system.scylla_local WHERE key = 'raft_group0_id'"))[0].value
+
+    rf: int = 3
+    ks_name = unique_name()
+    finish_writes = await start_writes(cql, rf, ConsistencyLevel.LOCAL_QUORUM, 5, ks_name, True)
```

Comment:
Nit: you could use named parameters for the arguments, especially for the concurrency (5) and ignore_errors (`True`).